### PR TITLE
introduce chunkserver watchdog

### PR DESCRIPTION
**Pull request by qinyihust**
### What problem does this PR solve?
detect file/device error for chunkserver, to warning or handle error in early time. 

How it Works:
run watchdog in background inside each chunkserver process
A series of watcher is running for error detection,
if any watcher detect error, print error log or kill the process.

Side effects(Breaking backward compatibility? Performance regression?): 
none

### Check List
- [ yes ] Relevant documentation/comments is changed or added
- [ yes ] I acknowledge that all my contributions will be made under the project's license

### CODE REVIEW

File: tools/watchdog/watcher.cpp [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+
+#include <fstream>
+#include <memory>
+#include <string>
+
+namespace Curve {
+namespace WatchDog {
+
+int ThreadWatcher::Run() {
+    char status = 0;
```
Comment (wu-hanqing) [inline tools/watchdog/watcher.cpp:39]: 这里status是不是可以封装成一个enum？

File: tools/watchdog/watcher.cpp [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+                       << getServerName(info->serverType) << " " << info->pid
+                       << ", failed to open device " << info->device;
+            return -1;
+        }
+        disk.close();
+    }
+
+    {
+        /* test SMART */
+        std::string cmd =
```
Comment (wu-hanqing) [inline tools/watchdog/watcher.cpp:89]: 这个命令执行，会不会卡磁盘的IO？ceph之前好像出现过smartctl卡IO的情况

File: tools/watchdog/watcher.cpp [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+        int ret = Exec(cmd, &output);
+        if (ret < 0) {
+            LOG_EVERY_N(ERROR, WATCHDOG_DUP_LOG_FREQUENCY)
+                << "SMART check ignored for " << info->device;
+            return 0;
+        }
+
+        if (output.empty()) {
+            /* print entire smartcl output for debug */
+            cmd = "smartctl -s on -H " + info->device;
```
Comment (wu-hanqing) [inline tools/watchdog/watcher.cpp:101]: 同上？

File: tools/watchdog/watchdog_main.cpp [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    "Usage: watchdog [OPTIONS...]\n"
+    "must to run with root privilege.\n"
+    "OPTIONS:\n"
+    "-scanCycleSec : time interval to scan disk\n"
+    "-opTimeoutSec : max time for a certain operation\n"
+    "-killTimeoutSec : max time for killing chunkserver\n"
+    "-logPath : log path for chunkserver_watchdog\n";
+
+static void print_config() {
+    LOG(INFO) << "chunkserver watchdog is starting...";
```
Comment (wu-hanqing) [inline tools/watchdog/watchdog_main.cpp:68]: 可以用一个LOG(INFO)

File: tools/watchdog/dog_keeper.cpp [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+
+    pthread_t pth;
+    if (pthread_create(&pth, NULL, Watch, this)) {
+        LOG(ERROR) << "failed to create watchdog thread for server "
+                   << serverInfo_->pid;
+        health_ = DogHealth::DOG_DEAD;
+        return -1;
+    }
+
+    tid_ = syscall(SYS_gettid);
```
Comment (wu-hanqing) [inline tools/watchdog/dog_keeper.cpp:88]: gettid也是一样的吧？

File: tools/watchdog/dog_keeper.cpp [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+int DogKeeper::AssignDogForServer() {
+    PROCTAB* proc = openproc(PROC_FILLCOM | PROC_FILLSTATUS);
+    proc_t procInfo = {0};
+    while (readproc(proc, &procInfo) != nullptr) {
+        /* filter server process */
+        ServerType serverType = GetServerType(&procInfo);
+        if (serverType == ServerType::UNKNOWN) continue;
+
+        /* skip zombie server process */
+        if (procInfo.state == 'Z') continue;
```
Comment (wu-hanqing) [inline tools/watchdog/dog_keeper.cpp:119]: 这里如果是Z，而且是chunkserver，是不是应该打个ERROR？

File: tools/watchdog/dog_keeper.cpp [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+
+        /* skip if process is already watched */
+        int pid = procInfo.tid;
+        auto iter = dogs_.find(pid);
```
Comment (wu-hanqing) [inline tools/watchdog/dog_keeper.cpp:123]: 可以用dogs_.count(pid) != 0来判断，而且之类多线程访问，是不是要加锁？

File: tools/watchdog/dog_keeper.cpp [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+        if (serverType == ServerType::CHUNKSERVER) {
+            std::shared_ptr<ChunkServerWatchDog> dog =
+                std::make_shared<ChunkServerWatchDog>(conf_, procInfo,
+                                                      serverType);
+            if (dog->Init() < 0) {
+                LOG(ERROR) << "failed to init dog for server " << pid;
+                closeproc(proc);
+                return -1;
+            }
+            dogs_.emplace(pid, dog);
```
Comment (wu-hanqing) [inline tools/watchdog/dog_keeper.cpp:136]: 这里需要加锁吗？

File: tools/watchdog/dog_keeper.h [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+            /* food is eaten up, fill it */
+            food_ = 1;
+        } else {
+            LOG(ERROR) << "food not eaten ,dog for server " << serverInfo_->pid
+                       << " is sick";
+            MarkSick();
+        }
+    }
+
+    bool IsBad() {
```
Comment (wu-hanqing) [inline tools/watchdog/dog_keeper.h:84]: 这种不修改成员变量的函数，可以加上const，`bool IsBad() const`

File: tools/watchdog/dog_keeper.cpp (hunk 7 above)
Comment (qinyihust) [reply tools/watchdog/dog_keeper.cpp:123]: dog keeper只有一个线程，dogs_不涉及多线程访问

File: tools/watchdog/dog_keeper.cpp (hunk 8 above)
Comment (qinyihust) [reply tools/watchdog/dog_keeper.cpp:136]: dogs_是单线程访问，不用加锁

File: tools/watchdog/dog_keeper.h (hunk 9 above)
Comment (qinyihust) [reply tools/watchdog/dog_keeper.h:84]: done

File: tools/watchdog/watchdog_common.cpp [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+    fl.l_whence = SEEK_SET;
+    fl.l_len = 0;
+    return fcntl(fd, F_SETLK, &fl);
+}
+
+int lockPidFile() {
+    int fd;
+    fd = open(WATCHDOG_LOCKFILE, O_RDWR | O_CREAT, WATCHDOG_LOCKMODE);
+    if (fd < 0) {
+        std::cerr << "can't open " << WATCHDOG_LOCKFILE << ": "
```
[the remark this answers is not in the document]
Comment (qinyihust) [reply tools/watchdog/watchdog_common.cpp:110]: done

File: tools/watchdog/watchdog_common.cpp [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+        return -1;
+    }
+
+    string line;
+    while (getline(procFile, line)) {
+        std::size_t pos = line.find("State:");
+        if (pos == string::npos) continue;
+
+        /* found */
+        line.erase(0, pos + sizeof("State:"));
```
[the remark this answers is not in the document]
Comment (qinyihust) [reply tools/watchdog/watchdog_common.cpp:78]: 已改为strlen

File: tools/watchdog/watchdog_common.cpp [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+
+namespace Curve {
+namespace WatchDog {
+
+using std::string;
+
+int checkEnv() {
+    int status = system("smartctl -h > /dev/null");
+    if (status) {
+        std::cerr << "Error, smartctl is not available."
```
[the remark this answers is not in the document]
Comment (qinyihust) [reply tools/watchdog/watchdog_common.cpp:44]: done

File: tools/watchdog/chunkserver_watchdog.cpp [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+
+    /* search mount list */
+    struct mntent* m;
+    FILE* mountFile = setmntent("/proc/mounts", "r");
+    if (!mountFile) {
+        LOG(ERROR) << "error read /proc/mounts:" << strerror(errno);
+        return "";
+    }
+
+    string deviceName = "";
```
[the remark this answers is not in the document]
Comment (qinyihust) [reply tools/watchdog/chunkserver_watchdog.cpp:56]: done

File: tools/watchdog/chunkserver_watchdog.cpp [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+#include "tools/watchdog/watcher.h"
+
+namespace Curve {
+namespace WatchDog {
+
+using std::string;
+
+string ChunkServerWatchDog::pathToDeviceName(string* path) {
+    /* erase last '/' in path */
+    while (path->back() == '/') {
```
[the remark this answers is not in the document]
Comment (qinyihust) [reply tools/watchdog/chunkserver_watchdog.cpp:42]: done

File: tools/watchdog/dog_keeper.cpp (hunk 5 above)
Comment (qinyihust) [reply tools/watchdog/dog_keeper.cpp:88]: Note: There is no glibc wrapper for this system call

File: src/chunkserver/watchdog/watcher.h [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+#include <memory>
+
+#include "src/chunkserver/watchdog/common.h"
+#include "src/chunkserver/watchdog/helper.h"
+
+namespace curve {
+namespace chunkserver {
+class Watcher {
+ public:
+    Watcher(volatile DogHealth* health, WatcherName watcherName,
```
Comment (wu-hanqing) [inline src/chunkserver/watchdog/watcher.h:35]: it's better to receive a dog in here rather then is internal state `health`, and check health through dog->IsHealth()

File: src/chunkserver/watchdog/helper.cpp [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+#include <fcntl.h>
+#include <glog/logging.h>
+#include <sys/stat.h>
+
+#include "src/chunkserver/watchdog/common.h"
+
+namespace curve {
+namespace chunkserver {
+
+int WatchdogHelper::Exec(const string cmd, string* output) {
```
[the remark this answers is not in the document]
Comment (qinyihust) [reply src/chunkserver/watchdog/helper.cpp:35]: done

File: src/chunkserver/watchdog/common.cpp [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+#include <string>
+
+namespace curve {
+namespace chunkserver {
+
+using std::string;
+
+int checkEnv() {
+    if (system("smartctl -h > /dev/null")) {
+        LOG(ERROR) << "Error, smartctl is not available";
```
[the remark this answers is not in the document]
Comment (qinyihust) [reply src/chunkserver/watchdog/common.cpp:41]: done

File: src/chunkserver/watchdog/common.cpp [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+
+    return 0;
+}
+
+string pathToDeviceName(const string* storDir) {
+    if (storDir->empty()) return "";
+
+    /* erase last '/' in path */
+    string path = *storDir;
+    if (path.empty()) return "";
```
[the remark this answers is not in the document]
Comment (qinyihust) [reply src/chunkserver/watchdog/common.cpp:55]: done

File: src/chunkserver/watchdog/common.cpp [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+
+    while (path.back() == '/') {
```
[the remark this answers is not in the document]
Comment (qinyihust) [reply src/chunkserver/watchdog/common.cpp:57]: done

File: src/chunkserver/watchdog/common.cpp [hunk 20]
```diff
@@ -0,0 +1,84 @@
...
+#include <fstream>
+#include <iostream>
+#include <string>
+
+namespace curve {
+namespace chunkserver {
+
+using std::string;
+
+int checkEnv() {
```
[the remark this answers is not in the document]
Comment (qinyihust) [reply src/chunkserver/watchdog/common.cpp:39]: done

File: src/chunkserver/watchdog/common.h [hunk 21]
```diff
@@ -0,0 +1,156 @@
...
+    while (0)
+
+void FdCleaner(void* arg);
+
+class PipeGuard {
+ public:
+    PipeGuard() : fp_(nullptr) {}
+    explicit PipeGuard(FILE* fp) : fp_(fp) {}
+
+    ~PipeGuard() noexcept(false) {
```
[the remark this answers is not in the document]
Comment (qinyihust) [reply src/chunkserver/watchdog/common.h:101]: to fix bug for pthread_cancel, no use now

File: src/chunkserver/watchdog/common.h [hunk 22]
```diff
@@ -0,0 +1,156 @@
...
+#include <glog/logging.h>
+#include <unistd.h>
+
+#include <memory>
+#include <string>
+#include <vector>
+
+namespace curve {
+namespace chunkserver {
+#define WATCHDOG_DEF_ENABLE true
```
[the remark this answers is not in the document]
Comment (qinyihust) [reply src/chunkserver/watchdog/common.h:35]: done

File: src/chunkserver/watchdog/helper.cpp [hunk 23]
```diff
@@ -0,0 +1,80 @@
...
+            << "Failed to open pipe for cmd: " << cmd << ", "
+            << strerror(errno);
+        return -1;
+    }
+
+    output->clear();
+    char buffer[WATCHER_EXEC_BUFFER_LEN];
+    while (!feof(pipe)) {
+        if (fgets(buffer, WATCHER_EXEC_BUFFER_LEN, pipe) != NULL) {
+            if (output->size() + strnlen(buffer, WATCHER_EXEC_BUFFER_LEN) >=
```
[the remark this answers is not in the document]
Comment (qinyihust) [reply src/chunkserver/watchdog/helper.cpp:48]: removed

File: src/chunkserver/watchdog/watcher.cpp [hunk 24]
```diff
@@ -0,0 +1,261 @@
...
+    fd =
+        helper_->Open(testPath.c_str(), O_CREAT | O_RDWR | O_NOATIME | O_DSYNC);
+    if (fd < 0) {
+        LOG(ERROR) << "Failed to create test file: " << testPath << ", "
+                   << strerror(errno);
+        return -1;
+    }
+
+    /* write file */
+    ret = WriteData(fd, writeBuf.c_str(), 0, WATCHDOG_WRITE_BUF_LEN);
```
[the remark this answers is not in the document]
Comment (qinyihust) [reply src/chunkserver/watchdog/watcher.cpp:215]: done

File: src/chunkserver/watchdog/watcher.h (hunk 15 above)
Comment (qinyihust) [reply src/chunkserver/watchdog/watcher.h]: done

**Comment by wu-hanqing** [comment]
recheck

**Comment by wu-hanqing** [comment]
recheck

**Comment by qinyihust** [comment]
recheck

**Comment by qinyihust** [comment]
recheck

**Comment by qinyihust** [comment]
recheck

File: conf/chunkserver.conf.example [hunk 25]
```diff
@@ -192,9 +192,34 @@ trash.expire_afterSec=300
 # chunkserver检查回收数据过期时间的周期
 trash.scan_periodSec=120
 
+#
+# watchdog settings
```
Comment (ilixiaocui) [inline conf/chunkserver.conf.example:196]: tile: introduce chunkserver watchdog 

module name: commit-message

File: src/chunkserver/chunkserver.cpp [hunk 26]
```diff
@@ -20,62 +20,111 @@
  * Author: lixiaocui
  */
 
-#include <glog/logging.h>
+#include "src/chunkserver/chunkserver.h"
```
Comment (ilixiaocui) [inline src/chunkserver/chunkserver.cpp:23]: Pay attention to the following order

File: src/chunkserver/chunkserver.cpp [hunk 27]
```diff
@@ -122,26 +171,37 @@ int ChunkServer::Run(int argc, char** argv) {
...
+        WatchConf watchdogOptions;
+        enableWatchdog = InitWatchdogOptions(&conf, &watchdogOptions);
+        if (enableWatchdog) {
+            std::shared_ptr<WatchdogHelper> helper =
+                std::make_shared<WatchdogHelper>();
+            std::shared_ptr<Dog> dog = std::make_shared<Dog>(helper);
+            dogKeeper = std::make_shared<DogKeeper>(dog, watchdogOptions);
+            if (dogKeeper->Init() < 0) enableWatchdog = false;
+        }
+    }
```
Comment (ilixiaocui) [inline src/chunkserver/chunkserver.cpp:192]: else print error

File: src/chunkserver/chunkserver.cpp [hunk 28]
```diff
@@ -291,38 +349,35 @@ int ChunkServer::Run(int argc, char** argv) {
...
-    ret = server.AddService(&braftCliService,
-                        brpc::SERVER_DOESNT_OWN_SERVICE);
+    ret = server.AddService(&braftCliService, brpc::SERVER_DOESNT_OWN_SERVICE);
     CHECK(0 == ret) << "Fail to add BRaftCliService";
 
     // braftclient service
     BRaftCliServiceImpl2 braftCliService2;
-    ret = server.AddService(&braftCliService2,
-                        brpc::SERVER_DOESNT_OWN_SERVICE);
+    ret = server.AddService(&braftCliService2, brpc::SERVER_DOESNT_OWN_SERVICE);
```
Comment (ilixiaocui) [inline src/chunkserver/chunkserver.cpp:365]: Do not format existing files ！

File: src/chunkserver/watchdog/common.h [hunk 29]
```diff
@@ -0,0 +1,156 @@
...
+#define WATCHER_EXEC_BUFFER_LEN 4096
+#define WATCHER_EXEC_MAX_BUF_LEN 16384
+#define WATCHDOG_PATH_LEN 1024
+#define WATCHDOG_CMD_BUF_LEN 4096
+#define WATCHDOG_WRITE_BUF_LEN 4096
+#define WATCHDOG_FILE_RETRY_TIMES 3
+
+using std::string;
+
+enum class DogHealth {
```
[the remark this answers is not in the document]
Comment (ilixiaocui) [reply src/chunkserver/watchdog/common.h:66]: DogState?

File: src/chunkserver/watchdog/dog.cpp [hunk 30]
```diff
@@ -0,0 +1,179 @@
...
+#include <iostream>
+#include <memory>
+#include <thread>
+
+#include "src/chunkserver/watchdog/watcher.h"
+
+namespace curve {
+namespace chunkserver {
+
+static void* Watch(void* currDog) {
```
Comment (ilixiaocui) [inline src/chunkserver/watchdog/dog.cpp:38]: why not 
static void Watch(void* currDog)

File: src/chunkserver/watchdog/dog.cpp [hunk 31]
```diff
@@ -0,0 +1,179 @@
...
+    return 0;
+}
+
+int Dog::Run() {
+    if (!config_ || IsDead()) {
+        LOG(ERROR) << "unable to run watchdog, should init dog first";
+        return -1;
+    }
+
+    if (pthread_create(&pth_, NULL, Watch, this)) {
```
Comment (ilixiaocui) [inline src/chunkserver/watchdog/dog.cpp:103]: use std::thread?

File: src/chunkserver/watchdog/dog.cpp [hunk 32]
```diff
@@ -0,0 +1,176 @@
...
+
+void Dog::OnError() {
+    LOG_EVERY_N(ERROR, WATCHDOG_DUP_LOG_FREQUENCY)
+        << "Error detected by watchdog, storDir=" << config_->storDir;
+    SetDead();
+    if (config_->watchdogKillOnErr) {
+        SetDestroy();
+
+        /* gracefully kill chunkserver */
+        raise(SIGTERM);
```
[the remark this answers is not in the document]
Comment (ilixiaocui) [reply src/chunkserver/watchdog/dog.cpp:171]: Has it been actually tested?

File: src/chunkserver/watchdog/dog_keeper.cpp [hunk 33]
```diff
@@ -0,0 +1,154 @@
...
+                FLAGS_watchdogRevive = false;
+                keeper->ReviveDog();
+                LOG(INFO) << "watchdog revived, storDir="
+                          << keeper->conf_.storDir;
+            }
+            std::this_thread::sleep_for(
+                std::chrono::milliseconds(kWatchdogSleepMs));
+        } else {
+            keeper->FeedDog();
+            /* sleep 1 second */
```
Comment (ilixiaocui) [inline src/chunkserver/watchdog/dog_keeper.cpp:72]: Inconsistent comments

Comment (qinyihust) [reply src/chunkserver/watchdog/dog_keeper.cpp:72]: in fact, kWatchdogSleepFreq * kWatchdogSleepMs = 1000ms

File: src/chunkserver/watchdog/dog.h [hunk 34]
```diff
@@ -0,0 +1,113 @@
...
+    const WatchConf* config_;
+    std::vector<std::shared_ptr<Watcher>> watchers_;
+
+ private:
+    std::shared_ptr<WatchdogHelper> helper_;
+
+    /*
+     * if error detected, dog will be set health to DOG_SICK
+     */
+    volatile DogHealth health_;
```
[the remark this answers is not in the document]
Comment (qinyihust) [reply src/chunkserver/watchdog/dog.h:98]: volatile can ensure atomic access to basic data structure

File: conf/chunkserver.conf.example (hunk 25 above)
Comment (qinyihust) [reply conf/chunkserver.conf.example:196]: done

File: conf/chunkserver.conf.example [hunk 35]
```diff
@@ -192,9 +192,34 @@ trash.expire_afterSec=300
...
+#
+# 启用watchdog
```
[the remark this answers is not in the document]
Comment (qinyihust) [reply conf/chunkserver.conf.example]: done

File: conf/chunkserver.conf.example [hunk 36]
```diff
@@ -192,9 +192,34 @@ trash.expire_afterSec=300
...
+# watchdog判定IO时延超标的阈值
+watchdog.latency_excess_ms=16
+# watchdog判定IO时延异常的阈值
+watchdog.latency_abnormal_ms=64
+# watchdog判定慢盘的异常时延比例
+watchdog.latency_window_slow_ratio=20
+# watchdog检测到故障后是否kill chunkserver进程
+watchdog.kill_chunkserver_on_error=false
+# watchdog尝试kill进程后，在此时间后通过kill -9强制结束进程
+watchdog.kill_timeout_sec=30
```
[the remark this answers is not in the document]
Comment (qinyihust) [reply conf/chunkserver.conf.example:217]: done

File: src/chunkserver/chunkserver.cpp (hunk 26 above)
Comment (qinyihust) [reply src/chunkserver/chunkserver.cpp]: done

File: src/chunkserver/chunkserver.cpp (hunk 27 above)
Comment (qinyihust) [reply src/chunkserver/chunkserver.cpp:192]: done

File: src/chunkserver/chunkserver.cpp (hunk 28 above)
Comment (qinyihust) [reply src/chunkserver/chunkserver.cpp]: done

File: src/chunkserver/watchdog/dog.cpp (hunk 30 above)
Comment (qinyihust) [reply src/chunkserver/watchdog/dog.cpp]: done

File: src/chunkserver/watchdog/dog.cpp (hunk 31 above)
Comment (qinyihust) [reply src/chunkserver/watchdog/dog.cpp]: done

**Comment by ilixiaocui** [comment]
cc @qinyihust Please resubmit to the master branch if necessary.

**Pull request closed without merging**